modules/generator/processor/localblocks/query_range.go

```diff
@@ -0,0 +1,280 @@
+		"block":     m.BlockID,
+		"blockSize": m.Size,
+	})
+	defer span.Finish()
+
+	// Trim and align the request for this block. I.e. if the request is "Last Hour" we don't want to
+	// cache the response for that, we want only the few minutes time range for this block. This has
+	// size savings but the main thing is that the response is reuseable for any overlapping query.
+	req.Start, req.End = traceql.TrimToOverlap(req.Start, req.End, req.Step, uint64(m.StartTime.UnixNano()), uint64(m.EndTime.UnixNano()))
+
+	if req.Start >= req.End {
+		// After alignment there is no overlap or something else isn't right
+		return nil, nil
+	}
+
+	cached, name, err := p.queryRangeCacheGet(ctx, m, req)
+	if err != nil {
+		return nil, err
+	}
+
+	span.SetTag("cached", cached != nil)
+
+	if cached != nil {
+		return cached.Series, nil
+	}
+
+	// Not in cache or not cacheable, so execute
+	eval, err := traceql.NewEngine().CompileMetricsQueryRange(&req, false, timeOverlapCutoff, unsafe)
+	if err != nil {
+		return nil, err
+	}
+	f := traceql.NewSpansetFetcherWrapper(func(ctx context.Context, req traceql.FetchSpansRequest) (traceql.FetchSpansResponse, error) {
+		return b.Fetch(ctx, req, common.DefaultSearchOptions())
+	})
+	err = eval.Do(ctx, f, uint64(m.StartTime.UnixNano()), uint64(m.EndTime.UnixNano()))
+	if err != nil {
+		return nil, err
+	}
+
+	results := eval.Results().ToProto(&req)
+
+	if name != "" {
+		err = p.queryRangeCacheSet(ctx, m, name, &tempopb.QueryRangeResponse{
+			Series: results,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("writing local query cache: %w", err)
+		}
+	}
+
+	return results, nil
+}
+
+func (p *Processor) queryRangeCacheGet(ctx context.Context, m *backend.BlockMeta, req tempopb.QueryRangeRequest) (*tempopb.QueryRangeResponse, string, error) {
+	cacheable := req.Start <= uint64(m.StartTime.UnixNano()) && req.End >= uint64(m.EndTime.UnixNano())
+	if !cacheable {
+		return nil, "", nil
+	}
+
+	hash := queryRangeHashForBlock(req)
+
+	name := fmt.Sprintf("cache_query_range_%v.buf", hash)
+
+	data, err := p.walR.Read(ctx, name, m.BlockID, m.TenantID, nil)
+	if err != nil {
+		if errors.Is(err, backend.ErrDoesNotExist) {
+			// Not cached, but return the name/keypath so it can be set after
+			return nil, name, nil
+		}
+		return nil, "", err
+	}
+
+	resp := &tempopb.QueryRangeResponse{}
+	err = proto.Unmarshal(data, resp)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return resp, name, nil
+}
+
+func (p *Processor) queryRangeCacheSet(ctx context.Context, m *backend.BlockMeta, name string, resp *tempopb.QueryRangeResponse) error {
+	data, err := proto.Marshal(resp)
+	if err != nil {
+		return err
+	}
+
+	return p.walW.Write(ctx, name, m.BlockID, m.TenantID, data, nil)
+}
+
+func queryRangeHashForBlock(req tempopb.QueryRangeRequest) uint64 {
+	h := fnv.New64a()
+	buf := make([]byte, 8)
+
+	h.Write([]byte(req.Query))
+
+	binary.BigEndian.PutUint64(buf, req.Step)
```

Comment:
Why use this and not the native `fnv1a.AddUint64`? `binary.BigEndian.PutUint64` is only putting the int into the hash. Unsure if there's a big difference.

Reply:
Good question, was just following practice from other code like trace combiner. Will take a look to see if there is any meaningful difference.

Reply:
They are identical and agree this new package is more efficient. Updated.